proposals/ideas/xxx-flashflow.txt

```diff
@@ -0,0 +1,950 @@
+
+X.X Pre-Measurement Handshaking/Starting a Measurement
+
+Measurers start connecting to the target relay and each sends it a MSM_PARAMS
+cell. If the target is unwilling to be measured at this time, it responds with
+an error cell and closes the connection.  Otherwise it checks that the
+MSM_PARAMS cell indicates that the measurer is probably allowed to perform a
+measurement and that the parameters of the measurement are acceptable (e.g.
+the length isn't too long, the indicated relay identity is its own, etc.).
+Initially, this will always be true. If the target is happy, it sends a
+MSM_PARAMS_OK, otherwise it sends a MSM_ERR and closes the connection.
+
+Upon receiving a MSM_PARAMS_OK from the target, the measurer will open the
+remaining sockets with the target. If it receives a MSM_ERR, it reports the
+error to the coordinator and considers the measurement a failure.
+
+The payload of MSM_PARAMS cells [XXX more may need to be added]:
+
+  - num_socks    [2 bytes]
+  - msm_duration [1 byte]
+  - bg_conn      [1 byte]
+
+num_socks is the total number of sockets this measurer will open with the
+target for this measurement. msm_duration is the duration, in seconds, that
+the actual measurement will last.  bg_conn is a flag indicating whether or not
+this connection is to be used for background traffic reports (MSM_BG cells)
+instead of MSM_ECHO traffic.
+
+MSM_PARAMS_OK has no payload: it's just padding bytes to make the cell 514
+bytes long.
+
+The payload of MSM_ECHO cells:
+
+  - arbitrary bytes [max to fill up 514 byte cell]
+
+[XXX it's possible while implementing the verification of MSM_ECHO cells that
+it will be discovered that it will be impossible to allow the measurer to skip
+encryption of most cells while requiring the relay to decrypt all cells]
+
+The payload of MSM_ERR cells:
+
+  - err_code [1 byte]
+  - err_str  [possibly zero-len null-terminated string]
+
+The error code is one of:
+
+  [... XXX TODO ...]
+  255 -- OTHER
+
+The error string is optional in all cases. It isn't present if the first byte
+of err_str is null, otherwise it is present. It ends at the first null byte or
+the end of the cell, whichever comes first.
+
+X.X Measurement Mode
+
+The relay considers the measurement to have started the moment it receives the
+first MSM_ECHO cell from any measurer. At this point, the relay
+
+  - starts a repeating 1s timer on which it will report the amount of
+    background traffic on the appropriate connection
+  - enters "measurement mode" and limits the amount of background traffic it
+    handles according to the torrc option/consensus parameter
+
+The relay decrypts and echos back all MSM_ECHO cells it receives on
+measurement connections until it has reported its amount of background traffic
+the same number of times as there are seconds in the measurement (e.g. 30
+per-second reports for a 30 second measurement). After sending the last
+MSM_BG cell, the relay drops all buffered MSM_ECHO cells, closes all
```

Comment:
I've documented MSM_BG cells as follows. I don't think the cell needs more than a single way of indicating which second it is for.

```text
The payload of MSM_BG cells:

  - second        [1 byte]
  - sent_bg_bytes [4 bytes]
  - recv_bg_bytes [4 bytes]

second is the number of seconds since the measurement began. MSM_BG cells are
sent once per second. The first cell will have this set to 1, and each
subsequent cell will increment it by one. sent_bg_bytes is the number of
background traffic bytes sent in the last second (since the last MSM_BG cell).
recv_bg_bytes is the same but for received bytes.
```

Reply:
Additional discussion about the contents of various cells can happen in future PRs or whatever medium we choose. Marking this as "resolved".